tests/in/glsl/samplers.frag

```diff
@@ -1,12 +1,229 @@
+    float d;
+    d = texture(sampler2DShadow(tex2DShadow, sampShadow), vec3(coord, 1.0));
+    d = textureGrad(sampler2DShadow(tex2DShadow, sampShadow), vec3(coord, 1.0), vec2(4.0), vec2(4.0));
+    d = textureGradOffset(sampler2DShadow(tex2DShadow, sampShadow), vec3(coord, 1.0), vec2(4.0), vec2(4.0), ivec2(5));
+    // d = textureLod(sampler2DShadow(tex2DShadow, sampShadow), vec3(coord, 1.0), 3.0);
+    // d = textureLodOffset(sampler2DShadow(tex2DShadow, sampShadow), vec3(coord, 1.0), 3.0, ivec2(5));
+    d = textureOffset(sampler2DShadow(tex2DShadow, sampShadow), vec3(coord, 1.0), ivec2(5));
+    d = textureProj(sampler2DShadow(tex2DShadow, sampShadow), vec4(coord, 1.0, 6.0));
+    d = textureProjGrad(sampler2DShadow(tex2DShadow, sampShadow), vec4(coord, 1.0, 6.0), vec2(4.0), vec2(4.0));
+    d = textureProjGradOffset(sampler2DShadow(tex2DShadow, sampShadow), vec4(coord, 1.0, 6.0), vec2(4.0), vec2(4.0), ivec2(5));
+    // d = textureProjLod(sampler2DShadow(tex2DShadow, sampShadow), vec4(coord, 1.0, 6.0), 3.0);
+    // d = textureProjLodOffset(sampler2DShadow(tex2DShadow, sampShadow), vec4(coord, 1.0, 6.0), 3.0, ivec2(5));
+    d = textureProjOffset(sampler2DShadow(tex2DShadow, sampShadow), vec4(coord, 1.0, 6.0), ivec2(5));
+}
+
+void testTex2DArray(in vec3 coord) {
+    vec4 c;
+    c = texture(sampler2DArray(tex2DArray, samp), coord);
+    c = texture(sampler2DArray(tex2DArray, samp), coord, 2.0);
+    c = textureGrad(sampler2DArray(tex2DArray, samp), coord, vec2(4.0), vec2(4.0));
+    c = textureGradOffset(sampler2DArray(tex2DArray, samp), coord, vec2(4.0), vec2(4.0), ivec2(5));
+    c = textureLod(sampler2DArray(tex2DArray, samp), coord, 3.0);
+    c = textureLodOffset(sampler2DArray(tex2DArray, samp), coord, 3.0, ivec2(5));
+    c = textureOffset(sampler2DArray(tex2DArray, samp), coord, ivec2(5));
+    c = textureOffset(sampler2DArray(tex2DArray, samp), coord, ivec2(5), 2.0);
+}
+
+void testTex2DArrayShadow(in vec3 coord) {
+    float d;
+    d = texture(sampler2DArrayShadow(tex2DArrayShadow, sampShadow), vec4(coord, 1.0));
+    d = textureGrad(sampler2DArrayShadow(tex2DArrayShadow, sampShadow), vec4(coord, 1.0), vec2(4.0), vec2(4.0));
+    d = textureGradOffset(sampler2DArrayShadow(tex2DArrayShadow, sampShadow), vec4(coord, 1.0), vec2(4.0), vec2(4.0), ivec2(5));
+    // d = textureLod(sampler2DArrayShadow(tex2DArrayShadow, sampShadow), vec4(coord, 1.0), 3.0);
+    // d = textureLodOffset(sampler2DArrayShadow(tex2DArrayShadow, sampShadow), vec4(coord, 1.0), 3.0, ivec2(5));
+    d = textureOffset(sampler2DArrayShadow(tex2DArrayShadow, sampShadow), vec4(coord, 1.0), ivec2(5));
+}
+
+void testTexCube(in vec3 coord) {
+    vec4 c;
+    c = texture(samplerCube(texCube, samp), coord);
+    c = texture(samplerCube(texCube, samp), coord, 2.0);
+    c = textureGrad(samplerCube(texCube, samp), coord, vec3(4.0), vec3(4.0));
+    c = textureGradOffset(samplerCube(texCube, samp), coord, vec3(4.0), vec3(4.0), ivec3(5));
```

Comment:
This isn't allowed by the glsl spec and `glslang` will yell at you, `textureGradOffset` doesn't support `Cube` textures.

Reply:
That's a fun exception! lol